CIP-0062? | Cardano dApp-Wallet Web Bridge Catalyst Extension (#296)
cardano-foundation/CIPs: CIP-0062/README.md

  weight: number
}

```

`votingKey`: Ed25519 pubkey 32 bytes HEX string  

`weight`: Used to calculate the actual voting power using the rules described
in 
[CIP-36](https://cips.cardano.org/cips/cip36/).

### **Purpose**

```
type enum Purpose = {
  CATALYST = 0,
  OTHER = 1
}

```

`Purpose`: Defines the purpose of the delegations. This is used to limit the scope of the delegations.  For example, a purpose might be a subset of Catalyst proposals, a council election, or even some private purpose (agreed by convention).

## **Namespace**

### **cardano.{walletName}.governance.enable(): Promise\<API>**
The `cardano.{walletName}.governance.enable()` method is used to enable the governance API. It should request permission from the wallet to enable the API. If permission is granted, the rest of the API will be available. The wallet should maintain a specific whitelist of allowed clients for this API. This whitelist can be used to avoid asking for permission every time.

This api being an extension of [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/), expects that `cardano.{walletName}.enable()` to be enabled and added to CIP-30 whitelist implicitly.

> When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

Author: That's correct @refi93 . I've updated this section, could you have a re-read?

> When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

Reviewer: I'm checking https://github.com/cardano-foundation/CIPs/pull/296/files#diff-d3ac75d45e4e621f6ad95567378e0dc2b671f44d3c32425dda9ca08210e18876R265-R272 but I didn't notice a difference, am I checking perhaps the wrong commit/branch?

> When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

Author: > Doesn't this sentence imply that the dapp has to **always** enable first the CIP-30 API and only then request the governance one? On the other hand this sentence

Yes, it ties to explain that CIP-30 should be enabled as well, if not so already. 

Maybe Is there a reason why it shouldn't? 

> Is my understanding that CIP-30 always has to be enabled before CIP-62 correct?

Yes, that was the idea. 

Would it be more clear if i were to remove the section that states: 
```
When both this API and CIP-30 is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.
```

# **`Jormungandr API`**

> ## **api.submitVotes**(votes: Vote[], spendingCounter: number): Promise\<hash32>

Reviewer: It seems the responsibility of this call is to construct a signed vote transaction containing the votes, using the currently registered voting key and submit it, am I right? I'd clarify this with a comment. Also, if there isn't any voting key registered, I guess this call should throw some error as the vote wouldn't be valid but I see no error code for this.

### `spendingCounter`: 
The spending counter is used to prevent double voting. The current spending counter for the account should be provided and the implementation should increment it for each vote before submission and attach it to the vote according to [Jormungandr Voting] (https://input-output-hk.github.io/jormungandr/jcli/vote.html#voting). This needs to be in sequential order.

### **`votes`**: 

```
interface Vote {
  proposal: Proposal,
  choice: number,
  expiration: BlockDate
}
```
#### `expiration`: 
chain epoch \& slot for when the vote will expire. The type used is: 

```
interface BlockDate {
    epoch: number
    slot: number
}
```

#### `choice`: 
The choice **index** we want to vote for. An `UnkownChoiceError` should be thrown is the value is not within the `proposal` option set.

#### **`proposal`** : 
proposal information. Include the range of options we can use to vote. This defines the allowed values in `choice`.

```

> interface Proposal {

Reviewer: will there be an API where the wallet could fetch the proposal metadata to show something more user friendly in the confirmation prompt to the user?

Author: I see; this would definitely be easier to implement on the client side. 

About some sort of Map<string,string> // <planId, proposalTitle> ?

    purpose: Purpose
}
```

Defines the structure to be crafted and signed for delegation of voting & their respectively voting power.   Embeds the stake key and reward address from the wallet, and constructs a suitable nonce.

***`delegations`***: List of keys and their voting weight to delegate voting power to.

This should be a call that implicitly cbor encodes the `delegation` object and uses the already existing [CIP-30](https://cips.cardano.org/cips/cip30/) `api.submitTx` to submit the transaction. The resulting transaction hash should be returned.

This should be trigger a request to the wallet to approve the transaction.

### **Returns**

```
interface SignedDelegationMetadata {
  certificate: DelegatedCertificate,
  signature: string, // signature on the CIP-36 delegation certificate
  txHash: string // of the transaction that submitted the delegation 
}
```

#### **DelegatedCertificate**
```
interface DelegatedCertificate {
  delegations: GovernanceKey[] // key delegations,
  stakingPub: string // staking public key
  rewardAddress: string // reward address
  nonce: number //nonce used in the transaction
  purpose: Purpose
}
```

Errors: `APIError`, `TxSendError`

> ## **Delegation Cert process**

Reviewer: I would expect a similar breakdown for the Voting process, it would make the motivation behind the available calls clearer to the reader

Reviewer 2: Will do.

1. **`Get Voting Key`** - use the method **api.getVotingKey** to return a ed25519 32 bytes public key (x value of the point on the curve).

2. **`Collect Voting Keys`** - Collect the keys to delegate voting power to.

> 3. **`Submit delegation`** - Submit the metadata transaction to the chain using **api.submitDelegation** which implicitly sign and/or can use the already existing **api.submitTx**,  available from [CIP-30](https://cips.cardano.org/cips/cip30/)

Reviewer: ```suggestion
3. **`Submit delegation`** - Submit the metadata transaction to the chain using **api.submitDelegation** which implicitly signs it
```
I think with the current CIP-0062 API it's not really possible/reasonable to use `api.submitTx` to submit the delegation transaction. The `api.submitDelegation` call doesn't return back the full signed registration transaction and even if it did, it seems counterproductive to submit the transaction again if the `submitDelegation` call is already required to do that, I'd even expect most implementation to await the tx being submitted before returning, making further submission by `api.submitTx` pointless

Author: Clearing this, as discussed that we only support signing votes through this API and the dapp should be able to submit it.

This api being an extension of [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/), expects that `cardano.{walletName}.enable()` to be enabled and added to CIP-30 whitelist implicitly.

When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

# **`Jormungandr API`**

## **api.submitVotes**(votes: Vote[], spendingCounter: number): Promise\<hash32>

### `spendingCounter`: 
The spending counter is used to prevent double voting. The current spending counter for the account should be provided and the implementation should increment it for each vote before submission and attach it to the vote according to [Jormungandr Voting] (https://input-output-hk.github.io/jormungandr/jcli/vote.html#voting). This needs to be in sequential order.

> 

Reviewer 3: How does the dApp get the "Current spending counter for the account"?

Reviewer 3: A: the dApp queries the catalyst node to get the current spending counter.

### **`votes`**: 

```
interface Vote {
  proposal: Proposal,
  choice: number,
  expiration: BlockDate
}
```
#### `expiration`: 
chain epoch \& slot for when the vote will expire. The type used is: 

```
interface BlockDate {
    epoch: number
    slot: number
}
```

> 

Reviewer 3: How does the dApp calculate the expiration BlockDate?  How does it know the current Jormungandr BlockDate in order to calculate an appropriate expiry?

`UnknownChoiceError` - If the `choice` is not within the `proposal` option set.

`InvalidBlockDateError` - If the `validUntil` is not a valid block date.

`InvalidVotePlanError` - If the `votePlanId` is not a valid vote plan.

`InvalidVoteOptionError` - If the `index` is not a valid vote option.

# **`Delegation API`**

## **api.getVotingKeys**(): Promise<cbor<PublicKey\>[]>
Should return a list of all the voting keys for the current wallet. 

### **Returns**
An array with the cbor hex encoded public keys.

## **api.rotateVotingKey**(): Promise<cbor<PublicKey\>>
This call should explicitly rotate the current in-use voting key. Given the current `address_index` in the derivation path defined in [CIP-36](https://cips.cardano.org/cips/cip36/), it should be incremented by 1.

The key should be derived from the following path. 

```
m / 1694' / 1815' / account' / role' / address_index'
```

`1694` (year Voltaire was born) Sets a dedicated `purpose` in the derivation path for the voting profile.  

`address_index` - index of the key to use. 

### **Returns**
cbor hex encoded representation of the public key

## **api.getCurrentVotingKey**(): Promise\<cbor<PublicKey\>>

> Should return the current in-use voting public-key. The wallet should maintain a reference to the current `adress_index` counter and return the public key for that index.

Author: clearing as this method doesn't exist anymore

### **Purpose**

```
type enum Purpose = {
  CATALYST = 0,
  OTHER = 1
}

```

`Purpose`: Defines the purpose of the delegations. This is used to limit the scope of the delegations.  For example, a purpose might be a subset of Catalyst proposals, a council election, or even some private purpose (agreed by convention).

## **Namespace**

### **cardano.{walletName}.governance.enable(): Promise\<API>**
The `cardano.{walletName}.governance.enable()` method is used to enable the governance API. It should request permission from the wallet to enable the API. If permission is granted, the rest of the API will be available. The wallet should maintain a specific whitelist of allowed clients for this API. This whitelist can be used to avoid asking for permission every time.

This api being an extension of [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/), expects that `cardano.{walletName}.enable()` to be enabled and added to CIP-30 whitelist implicitly.

When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

# **`Jormungandr API`**

> ## **api.submitVotes**(votes: Vote[], spendingCounter: number): Promise\<hash32>

Reviewer 4: IMHO method signature suggests that we are not using some Jormungandr node features:

- fail_fast mechanism - which can control what to do after nth fragment would be rejected ([example](https://github.com/input-output-hk/jormungandr/blob/master/testing/jormungandr-integration-tests/src/jormungandr/rest/v1/fragments/fail_fast.rs#L228))
- parallel_spending_counter - which allows to use 8 different lanes for spending counter. As a result each 8th fragments are independent from each other [example](https://github.com/input-output-hk/jormungandr/blob/master/testing/jormungandr-integration-tests/src/jormungandr/bft/counter.rs#L8)

is it omission?

Author: @dkijania this method was removed in favor of `signVotes`. Resolving. Please re-open if clarification is needed on the new method

These definitions extend [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/) to provide specific support for vote delegation.

# **Motivation**
The goal for this CIP is to extend the dApp-Wallet web bridge to enable the construction of transactions containing metadata that conforms to
[CIP-36 (Catalyst/Voltaire Registration Transaction Metadata Format - Updated)](https://cips.cardano.org/cips/cip36/),
enabling new functionality including vote delegation to either private or public representatives (dReps),
splitting or combining of private votes,
the use of different voting keys or delegations
for different purposes (Catalyst etc).

# **Specification**

## `Types`

### **GovernanceKey**

```
type GovernanceKey = {
  votingKey: string,
  weight: number
}

```

`votingKey`: Ed25519 pubkey 32 bytes HEX string  

`weight`: Used to calculate the actual voting power using the rules described
in 
[CIP-36](https://cips.cardano.org/cips/cip36/).

### **Purpose**

```
type enum Purpose = {
  CATALYST = 0,
  OTHER = 1

> }

Reviewer 5: Suggestion: reverse the order here. Keep `0` for other / misc, `1` for Catalyst and leave the door open for new purposes to be added after 1 :+1:

Reviewer 6: This reflects CIP-36.  0 was chosen for backwards compatibility with CIP-15.  The intention is indeed to use purposes from 1 onwards for various voting situations, in the same way that there is eg an RFC for port numbers, and a different RFC that defines the conventional uses of various ports.  At this point in time, we don't know what those purposes will be

Reviewer 3: I think `OTHER = 1` should be removed from the enum, its confusing as its not a real or defined current "Purpose".
There should just be a comment that says, other purposes may be defined in future revisions to this CIP.

Reviewer 7: I'd suggest finding a better type name, more self-explanatory than `Purpose.`

```

`Purpose`: Defines the purpose of the delegations. This is used to limit the scope of the delegations.  For example, a purpose might be a subset of Catalyst proposals, a council election, or even some private purpose (agreed by convention).

### **KeyPath**
```
interface KeyPath = {
  address_index: number 
  account: number
  role: number
}
```

## **Namespace**

### **cardano.{walletName}.governance.enable(): Promise\<API>**
The `cardano.{walletName}.governance.enable()` method is used to enable the governance API. It should request permission from the wallet to enable the API. If permission is granted, the rest of the API will be available. The wallet should maintain a specific whitelist of allowed clients for this API. This whitelist can be used to avoid asking for permission every time.

This api being an extension of [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/), expects that `cardano.{walletName}.enable()` to be enabled and added to CIP-30 whitelist implicitly.

When both this API and **CIP-30** is being enabled, is up to the wallet to decide the number of prompts requesting permissions to be displayed to the user.

> # **`Jormungandr API`**

Reviewer 5: Jörmungandr hasn't been mentioned so far, and comes a bit out of the blue here. I'd suggest to simply call this section "Governance API" or "API Extension"

Reviewer 8: resurrect the serpent

Reviewer 9: Agree. How about adding a `Purpose`arg to `api.submitVotes`? To me it seems like dapps just want to vote on Catalyst and Jormungandr is just an implementation detail that wallets can handle internally. Would probably need a new error type for unsupported voting purpose and/or a way to check if wallet supports it.

Reviewer 6: Yes, including the purpose makes sense

> ## **api.submitVotes**(keyPath: KeyPath, votes: Vote[], spendingCounter: number): Promise\<hash32>

Reviewer 5: I would recommend to split that function, or more specifically, to remove responsibilities from that function to keep the CIP-0030 API as orthogonal as possible. There are already is a function for submitting transactions (i.e. `submitTx`). Thus all that is really needed here is to:

- (a) create a transaction "skeleton" that contains the rightful metadata
- (b) sign that transaction with the appropriate signing keys (i.e. the stake keys)

I think both could be done at the same time, and return a partially signed transaction (for which input witnesses may still be provided later by `signTx`). So that the transaction can be inspected before submission.

}
```

#### `choice`: 
The choice **index** we want to vote for. An `UnkownChoiceError` should be thrown is the value is not within the `proposal` option set.

#### **`proposal`** : 
proposal information. Include the range of options we can use to vote. This defines the allowed values in `choice`.

```
interface Proposal {
  votePlanId: number
  voteOptions: number[]
}
```

##### `votePlanId`: 
the vote plan id. This is used to identify the vote plan. This should be the same as anchored on chain.

##### `voteOptions`: 
The vote options. This is the set of options we can vote for.

### **Returns**

`hash32` - This is the hash of the transaction that will be submitted to the node.

#### **Errors**
`InvalidArgumentError` - Generic error for errors in the formatting of the arguments.

`UnknownChoiceError` - If the `choice` is not within the `proposal` option set.

`InvalidBlockDateError` - If the `validUntil` is not a valid block date.

`InvalidVotingKeyError` - If the `keyPath` is not a valid voting key.

`InvalidVotePlanError` - If the `votePlanId` is not a valid vote plan.

`InvalidVoteOptionError` - If the `index` is not a valid vote option.

# **`Delegation API`**

> ## **api.getVotingKey**(path: KeyPath): Promise\<Bip32PublicKey>

Reviewer 5: This sounds like it could be a more generic API to get _any_ public key from the wallet.

Reviewer 6: You would need to specify what kind of key you wanted, but yes this could be extended

Author: @KtorZ i agree. A generic method to fetch this would warrant an update on CIP-30

Author: *keyPath* removed. Single Voting key generated now.

This should be trigger a request to the wallet to approve the transaction.

### **Returns**

```
interface SignedDelegationMetadata {
  certificate: DelegatedCertificate,
  signature: string, // signature on the CIP-36 delegation certificate
  txHash: string // of the transaction that submitted the delegation 
}
```

#### **DelegatedCertificate**
```
interface DelegatedCertificate {
  delegations: GovernanceKey[] // key delegations,
  stakingPub: string // staking public key
  rewardAddress: string // reward address
  nonce: number //nonce used in the transaction
  purpose: Purpose
}
```

Errors: `APIError`, `TxSendError`

## **Delegation Cert process**

1. **`Get Voting Key`** - use the method **api.getVotingKey** to return a ed25519 32 bytes public key (x value of the point on the curve).

2. **`Collect Voting Keys`** - Collect the keys to delegate voting power to.

3. **`Craft delegation cert`** - Use **api.buildDelegation** to construct the object containing the key array set to delegate voting power to. Each value will express the `weight` of the voting powers given.

> 4. **`Submit delegation`** - Submit the metadata transaction to the chain using **api.submitDelegation** which implicitly sign and/or can use the already existing **api.submitTx**,  available from [CIP-30](https://cips.cardano.org/cips/cip30/)

Reviewer 5: This proposal is missing a rationale section. As far as I can tell, a lot of those operations (like the creation of the metadata itself) can be achieved with client libraries such as the cardano-serialization-lib. Thus, what's the justification for adding them to the wallet connector?

Reviewer 5: (don't say it to me, write it in the proposal!)

This should trigger a request to the user of the wallet to approve the transaction.

##### Returns

The [Signed Delegation Metadata](#signeddelegationmetadata) of the voter registration delegation passed in the `delegation` parameter.

### Examples of Message Flows and Processes

#### Delegation

Recall from [CIP-36](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0036) a registration is a self-delegation, allocating one's voting power to one's own voting key.

1. **Get Voting Key** - dApp calls the method `api.getVotingCredentials()` to return the connected wallet account's public `voteKey`.

2. **Construct Delegation** - The dApp constructs `Delegation` using the Wallet's public `voteKey`, `weight` of 1 and choice of `VotingPurpose`.

3. **Submit Delegation** - The dApp passes the `Delegation` object to the Wallet to build a metadata transaction and submit this to Cardano blockchain. Wallets are able employ the already existing [`api.submitTx()`](https://cips.cardano.org/cips/cip30/#apisubmittxtxcbortransactionpromisehash32), available from [CIP-30](https://cips.cardano.org/cips/cip30/).

#### Voting

> TODO

Reviewer 10: todo?

## Rationale

To provide [CIP-36](https://cips.cardano.org/cips/cip36/) style governance specific functionality for Catalyst to wallet's and expose such API to the dApps (i.e Catalyst Voting Centers).

> This also addresses some short-comings of [CIP-30](https://cips.cardano.org/cips/cip30/); which signData can only be done by known an address; This signature is not relevant to a specific address, nor the dApp will know an address attached to the voting key. The voting key derivation is defined in [CIP-36](https://cips.cardano.org/cips/cip36/).

Reviewer 10: typo? "by known an address"

The implementation of this endpoint can make use of the already existing [CIP-30](https://cips.cardano.org/cips/cip30/) `api.signData` and `api.submitTx` to perform the broadcasting of the transaction containing the metadata.

Upon submission of the transaction containing the delegation cert as part of metadata, the wallet should store the delegation in its local storage and return an object that contains the delegation cert, the signature and the txhash of the transaction that the certificate was submitted with.

* delegation - the voter registration [delegation](#delegation) record.

This should be a call that implicitly cbor encodes the `delegation` object and uses the already existing [CIP-30](https://cips.cardano.org/cips/cip30/) `api.submitTx` to submit the transaction. The resulting transaction hash should be returned.

This should trigger a request to the user of the wallet to approve the transaction.

### Returns

The [Signed Delegation Metadata](#signeddelegationmetadata) of the voter registration delegation passed in the `delegation` parameter.

# Examples of Message Flows and Processes

## Delegation Cert process

> 1. **`Get Voting Key`** - dApp call the method [**api.getCurrentVotingKey**](#apigetcurrentvotingkey-promisecborpublickey) to return a ed25519 32 bytes public key (x value of the point on the curve).

Reviewer: I can't find the `api.getCurrentVotingKey` method in this doc anymore, was it dropped accidentally?

Author: Oops. Correct. fixing

---
CIP: 62
Title: Cardano dApp-Connector Governance extension
Authors: Bruno Martins <bruno.martins@iohk.io>, Steven Johnson <steven.johnson@iohk.io>
Status: Draft
Type: Standards
Created: 2021-06-11
License: CC-BY-4.0
---

# Abstract

> This document describe the interface between webpage / web-based stack and cardano wallets. This specifies that API of the javascript object that need to be injected into the web applications in order to support all the Governance features.

Author: Makes sense, will re-write :+1:

These definitions extend [CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/) to provide specific support for vote delegation.

# Motivation

The goal for this CIP is to extend the dApp-Wallet web bridge to enable the construction of transactions containing metadata that conforms to
[CIP-36 (Catalyst/Voltaire Registration Transaction Metadata Format - Updated)](https://cips.cardano.org/cips/cip36/),
enabling new functionality including vote delegation to either private or public representatives (dReps),
splitting or combining of private votes,
the use of different voting keys or delegations
for different purposes (Catalyst etc).

# Specification

## Version

The API Extension specified in this document will count as version 0.2.0 for version-checking purposes below.

## Data Types

### PublicKey

TODO: Define this.

### GovernanceKey

```js
type GovernanceKey = {
  votingKey: string,
  weight: number
}
```

* `votingKey` - Ed25519 pubkey 32 bytes HEX string
* `weight` - Used to calculate the actual voting power using the rules described
  in [CIP-36](https://cips.cardano.org/cips/cip36/).

### Voting Purpose

> ```js

Author: will update :+1:

type enum VotingPurpose = {
  CATALYST = 0
}
```

`VotingPurpose`: Defines the voting purpose of the governance functions. This is used
to limit the scope of the governance functions. For example, a voting purpose
might be a subset of Catalyst proposals, a council election, or even some
private purpose (agreed by convention). Currently, only voting purpose 0 (Zero)
is defined, and it is for Catalyst events.

**IMPORTANT**: The List of purposes in this CIP should be considered the
authoritative list of known purposes, subject to future amendment. Other voting
purposes will be defined as required, by either an update to this CIP or a
future CIP listing currently allocated Voting Purposes.

### BlockDate

```js
interface BlockDate {
  epoch: number
  slot: number
}
```

* epoch - an epoch value on the voting block chain.
* slot - a block slot number on the voting block chain.

#### Proposal

```js
interface Proposal {

>   votePlanId: number

Reviewer 3: This actually needs to be a hex string because the VotePlanId is a Hash of the voteplan, not a number.

Author: Makes sense. will update. Thanks!

  voteOptions: number[]
}
```

Proposal information.

* votePlanId - The unique ID of this proposal in the vote plan.  This will be the same as anchored on voting chain, and is managed by the dApp.
* voteOptions - List of possible "choices" which can be made for this proposal.

### Vote

```js
interface Vote {
  proposal: Proposal,
  choice: number,
  expiration: BlockDate
  purpose: VotingPurpose,
  spendingCounter: number

> }

Reviewer 3: Need to ensure the vote can be declared as either private/encrypted or public because signing the vote differs between them.

* rejectedVotes - (*OPTIONAL*) - In a voting transaction, there may be multiple votes being signed simultaneously.

### Extended TxSignError

```js
TxSignErrorCode {
  ProofGeneration: 1,
  UserDeclined: 2,
  VoteRejected: 3,
}
```

```js
type TxSignError = {
  code: TxSignErrorCode,
  info: String,
  rejectedVotes: number[]
}
```

All TxSignErrors defined in [CIP-30](https://cips.cardano.org/cips/cip30/#txsignerror) are unchanged.

* UserDeclined - Raised when the user declined to sign the entire transaction, in the case of a vote submission, this would be returned if the user declined to sign ALL of the votes.
* VoteRejected - On a vote transaction, where there may be multiple votes.  If the user accepted some votes, but rejected others, then this error is raised, AND `rejectedVotes` is present in the Error instance.

## Governance Extension to CIP-30

### cardano.{walletName}.governance.apiVersion: String

The version number of the Governance Extension API that the wallet supports.

> ### cardano.{walletName}.governance.enable(purpose: VotingPurpose[]): Promise\<API>

Author: This is to remove ambiguity of in-what context is the dapp requesting connection to the wallet. The prompt to the user should be clarify for application, use case is requesting connection. Otherwise after the whitelist, potentially the dApp could use it for other means. This was a separate method call, but through discussions and suggestion of @stevenj , it was included in the `enable` method.

}
```

* `votingKey` - Ed25519 pubkey 32 bytes HEX string
* `weight` - Used to calculate the actual voting power using the rules described
  in [CIP-36](https://cips.cardano.org/cips/cip36/).

### Voting Purpose

```js
type enum VotingPurpose = {
  CATALYST = 0
}
```

`VotingPurpose`: Defines the voting purpose of the governance functions. This is used
to limit the scope of the governance functions. For example, a voting purpose
might be a subset of Catalyst proposals, a council election, or even some
private purpose (agreed by convention). Currently, only voting purpose 0 (Zero)
is defined, and it is for Catalyst events.

**IMPORTANT**: The List of purposes in this CIP should be considered the
authoritative list of known purposes, subject to future amendment. Other voting
purposes will be defined as required, by either an update to this CIP or a
future CIP listing currently allocated Voting Purposes.

### BlockDate

```js
interface BlockDate {
  epoch: number
  slot: number
}
```

* epoch - an epoch value on the voting block chain.
* slot - a block slot number on the voting block chain.

#### Proposal

```js
interface Proposal {
  votePlanId: string
  voteOptions: number[]

>   votePublic: boolean

Reviewer: In the library to sign votes there's an additional proposalIndex parameter which I guess should be included here as well: https://github.com/input-output-hk/catalyst-core/blob/557808b7ceeeb35103abc2d7340b4be909b7ceb5/src/jormungandr/jcli/src/jcli_lib/certificate/new_vote_cast.rs#L22

Author: just added @refi93 :+1:

private purpose (agreed by convention). Currently, only voting purpose 0 (Zero)
is defined, and it is for Catalyst events.

**IMPORTANT**: The List of purposes in this CIP should be considered the
authoritative list of known purposes, subject to future amendment. Other voting
purposes will be defined as required, by either an update to this CIP or a
future CIP listing currently allocated Voting Purposes.

### BlockDate

```ts
interface BlockDate {
  epoch: number
  slot: number
}
```

* epoch - an epoch value on the voting block chain.
* slot - a block slot number on the voting block chain.

#### Proposal

```ts
interface Proposal {
  votePlanId: string
  voteOptions: number[]
  votePublic: boolean
  proposalIndex: number
}
```

Proposal information.

* votePlanId - Hex encoded string to represent vote plan unique identifier. This will be the same as anchored on voting chain, and is managed by the dApp.
* voteOptions - List of possible "choices" which can be made for this proposal.
* votePublic - Boolean indicating whether the vote is public or private.

### Vote

```ts
interface Vote {
  proposal: Proposal
  choice: number
  expiration: BlockDate
  purpose: VotingPurpose

>   spendingCounter: number

Reviewer: I think also wallet settings are missing here. The vote signer needs to know some basic network parameters to produce a valid vote. It is a stringified JSON in the vote signing lib: 

https://github.com/input-output-hk/catalyst-core/blob/main/src/chain-wallet-libs/bindings/wallet-js/js-test/test/vote_cast.js#L9

Reviewer: also, I'm not sure whether some parameter to manage the "lanes" wouldn't be needed in case of signing multiple votes with the `signVotes()` call: https://github.com/input-output-hk/catalyst-core/issues/131#issuecomment-1339195034

* VoteRejected - On a vote transaction, where there may be multiple votes.  If the user accepted some votes, but rejected others, then this error is raised, AND `rejectedVotes` is present in the Error instance.

## Governance Extension to CIP-30

### cardano.{walletName}.governance.apiVersion: String

The version number of the Governance Extension API that the wallet supports.

### cardano.{walletName}.governance.enable(purpose: VotingPurpose[]): Promise\<API>

Errors: [`APIError`](#extended-apierror)

The `cardano.{walletName}.governance.enable()` method is used to enable the
governance API. It should request permission from the wallet to enable the API for the requested purposes.

If permission is granted, the rest of the API will be available. The wallet
should maintain a specific whitelist of allowed clients and voting purposes for
this API. This whitelist can be used to avoid asking for permission every time.

This api, being an extension of
[CIP-30 (Cardano dApp-Wallet Web Bridge)](https://cips.cardano.org/cips/cip30/),

> expects that `cardano.{walletName}.enable()` to be enabled and added to CIP-30

Reviewer: ```suggestion
expects `cardano.{walletName}.enable()` to be enabled and added to CIP-30
```

* purpose - The [voting purpose](#voting-purpose) being voted on. (Currently always 0).
* spendingCounter - The spending counter is used to prevent double voting. The spending counter for the vote transaction will be supplied and maintained by the dApp. The dApp will manage supplying this in the correct order, and this should not be enforced by the Wallet.

It is required to attach it to the vote according to [Jormungandr Voting] (<https://input-output-hk.github.io/jormungandr/jcli/vote.html#voting>).

## Delegation

```ts
interface Delegation {
    delegations: GovernanceKey[]
    purpose: VotingPurpose
}
```

The record of a voters delegation.

* delegations - List of [Governance Keys](#governancekey) and their voting weight to delegate voting power to.
* purpose - The [Voting Purpose](#voting-purpose) being delegated.  A voter may have multiple active delegations for different purposes, but only 1 active delegation per unique Voting Purpose.

## DelegatedCertificate

```ts
interface DelegatedCertificate {
  delegations: GovernanceKey[]
  stakingPub: string

>   rewardAddress: string

Reviewer: I guess this should be renamed to match the latest CIP-36 specs: https://github.com/cardano-foundation/CIPs/pull/373

```suggestion
  paymentAddress: string
```

Reviewer 2: @refi93 

Yes good spot - I was going to reflect that change here once #373 is merged. 

Next week I will do a pre-emptive PR to this CIP to match the changes in #373 and #427 so when/if they are merged this CIP is aligned.

might be a subset of Catalyst proposals, a council election, or even some
private purpose (agreed by convention). Currently, only voting purpose 0 (Zero)
is defined, and it is for Catalyst events.

**IMPORTANT**: The List of purposes in this CIP should be considered the
authoritative list of known purposes, subject to future amendment. Other voting
purposes will be defined as required, by either an update to this CIP or a
future CIP listing currently allocated Voting Purposes.

### VotingCredentials
```ts
interface VotingCredentials {
  votingKey: PublicKey
  stakingCredential: PublicKey
}
```
Information used to represent a wallet's voting credentials.

* votingKey - Derivation as described within [CIP-0036](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0036), wallets should use `address_index`= 0.
* stakingCredential - At the moment, the only supported staking credential is a public staking key. This is used to represent the wallet's staked ADA.

### BlockDate

```ts
interface BlockDate {
  epoch: number
  slot: number
}
```

* epoch - an epoch value on the voting block chain.
* slot - a block slot number on the voting block chain.

#### Proposal

```ts

> interface Proposal {
>   votePlanId: string;
>   proposalIndex: number;
>   voteOptions?: number[];
>   voteEncKey?: string;
> }

Reviewer: since voteEncKey and voteOptions are required for private votes, I'd consider rewriting it like this to make that requirement explicit at type level:

```suggestion
interface Proposal =
  | {
      votePlanId: HexString
      voteOptions: number[]
      votePublic: false
      proposalIndex: number
      voteEncKey: HexString
    }
  | {
      votePlanId: HexString
      voteOptions: number[]
      votePublic: true
      proposalIndex: number
    }
```

Reviewer 2: This is sensible - I will make this change.

* proposalIndex  - The index of the proposal within the vote plan.
* voteOptions - *Optional*. Total number of vote options.  Only present in a private/encrypted vote.
* voteEncKey - *Optional*. If this vote is Private, this field is present and contains the key used to encrypt the vote.  When the vote is public, this key is absent from the proposal.

### Vote

```ts
interface Vote {
  proposal: Proposal;
  choice: number;
  expiration: BlockDate;
  purpose: VotingPurpose;
  spendingCounter?: number;
  spendingCounterLane?: number;
}
```

An individual raw Unsigned Vote record.

* proposal - The [proposal](#proposal) being voted on.
* choice - This value MUST match one of the available `voteOptions` in the `proposal` element. An `UnknownChoiceError` should be thrown if the value is not one of the valid `voteOptions` of the `proposal`.
* expiration - Voting Chain epoch \& slot for when the vote will expire. This value is supplied and maintained by the dApp, and forms a necessary component of a [Vote](#vote)
* purpose - The [voting purpose](#voting-purpose) being voted on. (Currently always 0).

> * spendingCounter - Optional, but if present, the spending counter is used to prevent double voting. The spending counter for the vote transaction will be supplied and maintained by the dApp. The dApp will manage supplying this in the correct order, and this should not be enforced by the Wallet.

Reviewer: if not provided, what value, if any should this default to?

Reviewer 2: These fields should default to 0 - thanks for raising this I will make it explicit in the specification.

a javascript object we will refer to as `API` (type) / `api` (instance) is
returned to the dApp with the following methods.

## Governance API

Except `signVotes`, no other method should require any user
interaction as the user has already consented to the dApp reading information
about the wallet's governance state when they agreed to
[`cardano.{walletName}.governance.enable()`](#cardanowalletnamegovernanceenablepurpose-votingpurpose-promiseapi).
The remaining methods
[`api.signVotes()`](#apisignvotesvotes-vote-promisebytes) and
[`api.signData()`](#apisubmitdelegationdelegation-delegation-promisesigneddelegationmetadata)
must request the user's consent in an informative way for each and every API
call in order to maintain security.

The API chosen here is for the minimum API necessary for dApp <-> Wallet
interactions without convenience functions that don't strictly need the wallet's
state to work.

## api.signVotes(votes: Vote[], settings: string): Promise\<Bytes>[]

Errors: [`APIError`](#extended-apierror), [`TxSignError`](#extended-txsignerror)

* `votes` - an array of up to 10 votes to be validated with the wallet user and, if valid, signed.

> * `settings` - *Optional*. Settings which are universally applicable to all `votes` being signed.  This is a json string.  The fields within the settings json string depend on the format of the vote record.  The wallet does not need to interpret this field and should pass it unchanged to the logic used to format the vote transaction.  However, to support multiple future vote transaction formats, the `settings` must contain at least:

Reviewer: if the wallet doesn't validate the purpose in settings, can't a malicious actor abuse that to pass a different purpose than the one whitelisted by the wallet? If so, I'd consider removing the purpose from the settings and passing it separately, or at least specifying that the wallet should validate that field in the settings

Reviewer: though looking at the js bidings lib to sign votes, I don't see the `purpose` field being used at all: https://github.com/input-output-hk/catalyst-core/blob/f5865be9a9d227e975689e32e6c03b5b34ff5d96/src/chain-wallet-libs/bindings/wallet-wasm-js/js-test/test/vote_cast.js#L10 - what's the benefit of passing `purpose` in the `settings` parameter then?

Reviewer 2: The intent of including the purpose here is to allow wallets to validate. Wallets should be able to show the supplied purpose of a vote to user at signing time for them to validate.

# **`Delegation API`**

> ## **api.getVotingKey**(): Promise\<cbor<PublicKey\>>

Reviewer: what should this call return if there's no voting key registered yet?

Reviewer: also, as per the CDDL, a user may register multiple voting keys in a single registration, and multiple of them may be theirs, in theory. In such case, I'd expect this call to return a list of voting keys, or is that case impossible/against specs?

Reviewer: > what should this call return if there's no voting key registered yet? I'd expect it to return null but the specs apparently don't mention that

ok depends on the definition of "in use", whether it means "registered" or "the next in line to be registered" or something else. But I'd expect to have a call that would be able to provide both keys - the latest registered voting key makes sense to be able to show voting stats/construct the voting transaction and the "the next in line to be registered" key makes sense for the dapp to build the voting delegation. I'd therefore expect actually two different calls to be able to retrieve any of these two keys, as needed, but maybe I'm missing something

Author: @refi93 The latest iteration of this spec, expects a single key to be created / "in-use" . We discussed supporting a N keys to be created, but for now seems unnecessary. Doesn't mean we wouldn't expand it in the future. 

Does it makes sense?

Reviewer: makes sense :+1:

## Governance API

> Except `signVotes`, no other method should require any user

Reviewer 2: `.submitDelegation()` should also need the user to authorise signing and submission of delegation transaction. Also, dependent on wallet implementations, `.getVotingKey()` may need user interaction -- to authorize mnemonic exposure for voting key derivation.
